### ISSUE TITLE
Variable functions and closures aren't supported

### ISSUE THREAD

**Issue by TazeTSchnitzel** [gosukiwi/Blueberry#29]
Blueberry doesn't support PHP's closures yet, which is a shame, because they're really cool. Perhaps something like the following could work:

    myClosure = def foo(bar) use (a, &b)
      return a + b + bar
    end

We could also clone Hack's `==>`, perhaps just making it `->` like CoffeeScript since we don't have the conflict PHP does. It auto-imports referenced variables from surrounding scope:

    myClosure = (bar) -> a + bar
    myClosure = (bar) -> do
      return a + b + bar
    end

The connected issue is variable functions. In PHP, you can do this:

    $x = 'strlen';
    $x('foobar');

But in Blueberry, it's currently not possible to do this, since there aren't any sigils. That's really quite a shame. I'm not sure how to solve this.

One approach would be, like CoffeeScript, to change how variables are scoped. If you had earlier declared a variable called `foobar` in your script, then we'd compile the Blueberry code `foobar()` to PHP `$foobar()`, but if you hadn't, we'd compile it to PHP `foobar()`. This would mean we'd have to explicitly declare variables global at the top level... which would probably be good thing! That, or get rid of them entirely.

Another approach is what Snowscript does, where all function declarations are done as variables. I really don't like that idea, though.

**Comment by gosukiwi** [comment]
I really like this syntax

```
myClosure = (bar) -> a + bar
myClosure = (bar) -> do
  return a + b + bar
end
```

**Comment by TazeTSchnitzel** [comment]
That works well :)

Main issue is the implicit variable capture.

**Comment by gosukiwi** [comment]
Yeah I think it's going to have to be

```
myClosure = (bar) use (a) -> a + bar
myClosure = (bar) use (a) -> do
  return a + b + bar
end
```

**Comment by TazeTSchnitzel** [comment]
Maybe.

**Comment by gosukiwi** [comment]
As for calling variable functions... Certainly `a()` would get compiled to `a();`. I'd say the simplest solution is add some syntax construct to know when it's supposed to be a regular function call and when a variable function call.

For example, `$a()` could get compiled to `$a();`. But adding the `$` character, might be confusing. Something I thought maybe was doing `call a()` and compile that to `$a();`.

**Comment by gosukiwi** [comment]
Okay I implemented it the hard way in 664073591ec5c334db676fb1b1fc649105e6af80. I added a very simple scope collector. So now we have access to all assigned variables, and also the ones in the parameters. Just a note: Class attributes cannot be closures, because it would be just silly.

**Comment by TazeTSchnitzel** [comment]
Awesome :)

Does this do implicit variable capture too? Because with the same infrastructure you can do that.

**Comment by gosukiwi** [comment]
You mean like adding the variables in the current scope to the closure's `use` automatically?

**Comment by TazeTSchnitzel** [comment]
No, just the ones you choose to use, e.g.:

    a = 3
    b = 2
    c = (z) -> z + b


    <?php
    $a = 3;
    $b = 2;
    $c = function (z) use ($b) { return $z + $b; }

This would work similarly to Hack's lambda functions which have the `==>` syntax:

http://docs.hhvm.com/manual/en/hack.lambda.examples.php

**Comment by gosukiwi** [comment]
Ah I see now. I'll give that a shot later today.

**Comment by TazeTSchnitzel** [comment]
Incidentally, it's something I want to add to PHP proper, but I'm too focussed on [certain](https://wiki.php.net/rfc/scalar_type_hints) [other](https://wiki.php.net/rfc/bigint) [things](https://github.com/php/php-src/pull/982) [just](https://github.com/php/php-src/pull/987) [now](http://news.php.net/php.internals/80383).

**Comment by gosukiwi** [comment]
Interesting! This was harder than I expected. Mostly because I have to parse a block, and actually see what variables that block requires. I implemented a dummy solution in a6d76a98506a39fb1c425b1ee1c1309a55d6135e. Will look at this closer eventually. For now I'd like to give the languages some other features such as static methods :)

**Comment by gosukiwi** [comment]
Moved discussion to #39
